File: sdk-js/src/utils/stream.utils.ts
import type {
  AudioPart,
  AudioPartDelta,
  ContentDelta,
  ModelResponse,
  TextPartDelta,
  ToolCallPart,
  ToolCallPartDelta,
} from "../types.js";
import {
  arrayBufferToBase64,
  base64ToArrayBuffer,
  mergeInt16Arrays,
} from "./audio.utils.js";

export type InternalAudioPartDelta = Omit<AudioPartDelta, "audioData"> & {
  audioData: ArrayBuffer[];
};

export interface InternalContentDelta {
  index: number;
  part: TextPartDelta | ToolCallPartDelta | InternalAudioPartDelta;
}

export class ContentDeltaAccumulator {
  deltas: InternalContentDelta[] = [];

  addChunks(incomingDeltas: ContentDelta[]) {
    for (const incomingDelta of incomingDeltas) {
      const existingDelta = this.deltas.find(
        (delta) => delta.index === incomingDelta.index,
      );

      if (existingDelta) {
        if (incomingDelta.part.id) {
          existingDelta.part.id = incomingDelta.part.id;
        }

        if (
          existingDelta.part.type === "text" &&
          incomingDelta.part.type === "text"
        ) {
          existingDelta.part.text += incomingDelta.part.text;
        } else if (
          existingDelta.part.type === "tool-call" &&
          incomingDelta.part.type === "tool-call"
        ) {
          if (incomingDelta.part.tool_name) {
            existingDelta.part.tool_name =
              (existingDelta.part.tool_name || "") +
              incomingDelta.part.tool_name;
          }
          if (incomingDelta.part.tool_call_id) {
            existingDelta.part.tool_call_id = incomingDelta.part.tool_call_id;
          }
          if (incomingDelta.part.args) {
            existingDelta.part.args =
              (existingDelta.part.args || "") + incomingDelta.part.args;
          }
        } else if (
          existingDelta.part.type === "audio" &&
          incomingDelta.part.type === "audio"
        ) {
          if (incomingDelta.part.audio_data) {
            const incomingAudioData = base64ToArrayBuffer(
              incomingDelta.part.audio_data,
            );
            // keep an array of audioBuffer internally and concat at the end
            existingDelta.part.audioData.push(incomingAudioData);
          }
          if (incomingDelta.part.format) {
            existingDelta.part.format = incomingDelta.part.format;
          }
          if (incomingDelta.part.sample_rate) {
            existingDelta.part.sample_rate = incomingDelta.part.sample_rate;
          }
          if (incomingDelta.part.channels) {
            existingDelta.part.channels = incomingDelta.part.channels;
          }
          if (incomingDelta.part.transcript) {
            existingDelta.part.transcript =
              (existingDelta.part.transcript || "") +
              incomingDelta.part.transcript;
          }
        } else {
          throw new Error(
            `unexpected part at index ${String(incomingDelta.index)}. existing part has type ${existingDelta.part.type}, incoming part has type ${incomingDelta.part.type}`,
          );
        }
      } else {
        this.deltas.push({
          index: incomingDelta.index,
          part: {
            ...(incomingDelta.part.type === "audio"
              ? {
                  ...incomingDelta.part,
                  audioData: incomingDelta.part.audio_data
                    ? [base64ToArrayBuffer(incomingDelta.part.audio_data)]
                    : [],
                }
              : incomingDelta.part),
          },
        });
      }
    }
  }

  computeContent(): ModelResponse["content"] {
    return this.deltas.map((delta): ModelResponse["content"][number] => {
      switch (delta.part.type) {
        case "text":
          return {
            ...(delta.part.id && { id: delta.part.id }),
            type: "text",
            text: delta.part.text,
          };
        case "tool-call":
          if (!delta.part.tool_call_id || !delta.part.tool_name) {
            throw new Error(
              `missing tool_call_id or tool_name at index ${String(delta.index)}. tool_call_id: ${String(delta.part.tool_call_id)}, tool_name: ${String(delta.part.tool_name)}`,
            );
          }
          return {
            type: "tool-call",
            ...(delta.part.id && { id: delta.part.id }),
            tool_call_id: delta.part.tool_call_id,
            args: delta.part.args
              ? (JSON.parse(delta.part.args) as Record<string, unknown>)
              : null,
            tool_name: delta.part.tool_name,
          };
        case "audio": {
          if (delta.part.format !== "linear16") {
            throw new Error(
              `only linear16 format is supported for audio concatenation. format: ${String(delta.part.format)}`,
            );
          }
          const concatenatedAudioData = mergeInt16Arrays(delta.part.audioData);
          const audioPart: AudioPart = {
            type: "audio",
            audio_data: arrayBufferToBase64(concatenatedAudioData),
          };
          if (delta.part.id) {
            audioPart.id = delta.part.id;
          }
          audioPart.format = delta.part.format;
          if (delta.part.sample_rate) {
            audioPart.sample_rate = delta.part.sample_rate;
          }
          if (delta.part.channels) {
            audioPart.channels = delta.part.channels;
          }
          if (delta.part.transcript) {
            audioPart.transcript = delta.part.transcript;
          }
          return audioPart;
        }
        default: {
          const exhaustiveCheck: never = delta.part;
          throw new Error(
            `unexpected part ${String(exhaustiveCheck)} at index ${String(delta.index)}`,
          );
        }
      }
    });
  }
}

// Because of difference in mapping, especially in
// openai cases, where text and audio part does not have indexes
// or google cases, where no parts have indexes
// we need to guess the index of the incoming delta
export function guessDeltaIndex(
  part: ContentDelta["part"] | ToolCallPart,
  allContentDeltas: (ContentDelta | InternalContentDelta)[],
  existingMatchingDelta?: ContentDelta | InternalContentDelta,
) {
  let matchingDelta = existingMatchingDelta;
  if (!matchingDelta) {
    matchingDelta = allContentDeltas.findLast((contentDelta) => {
      if (part.type === "text" || part.type === "audio") {
        return contentDelta.part.type === part.type;
      }
      // we won't be able to reliably match tool calls
      // because there can be multiple tool calls with the same tool name
      return false;
    });
  }
  if (matchingDelta) {
    return matchingDelta.index;
  }
  const maxIndex = Math.max(
    ...allContentDeltas.map((contentDelta) => contentDelta.index),
    -1,
  );
  return maxIndex + 1;
}

File: sdk-go/streamutils.go
package llmsdk

import "slices"

// GuessDeltaIndex tries to guess the appropriate index for a delta based on existing deltas.
//
// Because of the difference in mapping, especially in OpenAI cases,
// where text and audio part does not have indexes
// or in Google cases, where no parts have indexes,
// we need to guess an index for the incoming delta
// which is required in our unified interface.
//
// toolCallIndex does not always correspond to the index of the tool call in
// the deltas because some providers keep tool call separate from other parts
// (e.g openai). We can match this against the existing tool call deltas.
func GuessDeltaIndex(part PartDelta, allContentDeltas []ContentDelta, toolCallIndex *int) int {
	// contentDeltas may have the structure of
	// [part0 partial, part0 partial, part1 partial].
	// For the purpose of this matching, we want only
	// [part0, part1]
	uniqueContentDeltas := slices.CompactFunc(slices.Clone(allContentDeltas), func(a, b ContentDelta) bool {
		return a.Index == b.Index
	})

	if toolCallIndex != nil && part.ToolCallPartDelta != nil {
		// Providers like OpenAI track tool calls in a separate field, so we
		// need to reconcile that. To understand how this matching works:
		// [Provider]
		// toolCalls: [index 0] [index 1]
		// [LLM-SDK state]
		// parts: [index 0 text] [index 1 tool] [index 2 text] [index 3 tool]
		// In this case, we need to map the tool index 0 -> 1 and 1 -> 3
		var toolPartDeltas []ContentDelta
		for _, contentDelta := range uniqueContentDeltas {
			if contentDelta.Part.ToolCallPartDelta != nil {
				toolPartDeltas = append(toolPartDeltas, contentDelta)
			}
		}

		if *toolCallIndex < len(toolPartDeltas) {
			return toolPartDeltas[*toolCallIndex].Index
		}
		// If no matching tool call delta found, return the length of
		// uniqueContentDeltas. This is because we want to append a new tool
		// call delta
		return len(uniqueContentDeltas)
	}

	// Attempt to find the LAST matching delta in uniqueContentDeltas
	var matchingDelta *ContentDelta
	for i := len(uniqueContentDeltas) - 1; i >= 0; i-- {
		contentDelta := &uniqueContentDeltas[i]
		// Inline matching logic: For text and audio parts, they are the matching delta
		// if their types are the same. This is because providers that do not
		// provide indexes like only have 1 part for each type (e.g openai has only 1
		// message.content or 1 message.audio)
		isMatch := false
		if contentDelta.Part.TextPartDelta != nil && part.TextPartDelta != nil {
			isMatch = true
		} else if contentDelta.Part.AudioPartDelta != nil && part.AudioPartDelta != nil {
			isMatch = true
		}
		// For tool calls, we can't reliably match them
		// because there can be multiple tool calls with the same tool name
		// Different types don't match

		if isMatch {
			matchingDelta = contentDelta
			break
		}
	}

	if matchingDelta != nil {
		return matchingDelta.Index
	}

	// If no matching delta found, return max index + 1
	maxIndex := 0
	for _, contentDelta := range uniqueContentDeltas {
		if contentDelta.Index > maxIndex {
			maxIndex = contentDelta.Index
		}
	}

	// Since we're using int, we start from 0 instead of -1
	if maxIndex == 0 && len(uniqueContentDeltas) == 0 {
		return 0
	} else {
		return maxIndex + 1
	}
}

func LooselyConvertPartToPartDelta(part Part) PartDelta {
	switch {
	case part.TextPart != nil:
		return PartDelta{
			TextPartDelta: &TextPartDelta{
				Text: part.TextPart.Text,
			},
		}
	case part.ToolCallPart != nil:
		argsStr := string(part.ToolCallPart.Args)
		return PartDelta{
			ToolCallPartDelta: &ToolCallPartDelta{
				ToolCallID: &part.ToolCallPart.ToolCallID,
				ToolName:   &part.ToolCallPart.ToolName,
				Args:       &argsStr,
			},
		}
	case part.ReasoningPart != nil:
		return PartDelta{
			ReasoningPartDelta: &ReasoningPartDelta{
				Text:      part.ReasoningPart.Text,
				Signature: part.ReasoningPart.Signature,
			},
		}
	case part.ImagePart != nil:
		return PartDelta{
			ImagePartDelta: &ImagePartDelta{
				MimeType:  &part.ImagePart.MimeType,
				ImageData: &part.ImagePart.ImageData,
				Width:     part.ImagePart.Width,
				Height:    part.ImagePart.Height,
			},
		}
	case part.AudioPart != nil:
		return PartDelta{
			AudioPartDelta: &AudioPartDelta{
				AudioData:  &part.AudioPart.AudioData,
				Format:     &part.AudioPart.Format,
				SampleRate: part.AudioPart.SampleRate,
				Channels:   part.AudioPart.Channels,
				Transcript: part.AudioPart.Transcript,
				AudioID:    part.AudioPart.AudioID,
			},
		}
	default:
		return PartDelta{}
	}
}
